# Bioinformatics tutorial
This tutorial gives you an introduction to single cell RNA-seq data analysis.  
The contents include:
1. item1 how to manipulate scRNA-seq data with Python
2. item2 pre-process and visualization
3. item3 clustering and dimentinality reduction
4. item4 trajectory inference

Before beginning the tutorial, please set up your envirionment.  
1. item1 Install Python (> ver. 3.6)
2. item2 Install the following libraries  

*UNDER CONSTRUCTION*  

---  

## Key papers
### Data

* [A comprehensive single cell transcriptional landscape of human hematopoietic progenitors](https://www.nature.com/articles/s41467-019-10291-0)  

Pellin, D. et al., 2019, Nature Communications  

### Tool

* [SCANPY: large-scale single-cell gene expression data analysis](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0)

Wolf, F. A. et al., 2018, Genome Biology

* [PAGA: graph abstraction reconciles clustering with trajectory inference through a topology preserving map of single cells](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x)

Wolf, F. A. et al., 2019, Genome Biology  

* [Comparative single-cell trajectory network enrichment identifies pseudo-temporal systems biology patterns in hematopoiesis and CD8 T-cell development](https://www.biorxiv.org/content/10.1101/2020.04.02.021295v3)  
Gr√∏nning, A. G. B., et al., 2020, bioRxiv  

## Key websites

* [Scanpy tutorials](https://scanpy-tutorials.readthedocs.io/en/latest/index.html)

* [Ensembl](https://asia.ensembl.org/)  
* [Scellnetor](https://exbio.wzw.tum.de/scellnetor//)

